### PR TITLE
Add validation to document (and some tests)

### PR DESCRIPTION
**Pull request by timchurchard**

### CODE REVIEW

File: pkg/advancedapi/advancedApi_test.go [hunk 1]
```diff
@@ -7,7 +7,8 @@ import (
 	"testing"
 	"time"
 
-	advancedapi2 "github.com/Iotic-Labs/iotics-identity-go/pkg/advancedapi"
```
Comment (mirogta) [inline pkg/advancedapi/advancedApi_test.go:10]: yay 👍

File: pkg/register/document.go [hunk 2]
```diff
@@ -2,41 +2,56 @@
...
+	"strings"
+
+	"github.com/Iotic-Labs/iotics-identity-go/pkg/identity"
+	"github.com/Iotic-Labs/iotics-identity-go/pkg/validation"
+	"github.com/jbenet/go-base58"
+)
+
 const (
-	// DocumentContext context in document
+	// DocumentContext context in document.
```
Comment (mirogta) [inline pkg/register/document.go:16]: it's interesting how one can see who's and who's not using Goland 😄

File: pkg/register/documentBuilder.go [hunk 3]
```diff
@@ -62,10 +62,17 @@ func NewRegisterDocument(opts []RegisterDocumentOpts) (*RegisterDocument, []erro
...
 		return nil, errs
 	}
+
 	doc, err := builder.build()
 	if err != nil {
-		return nil, []error{err}
+		errs = append(errs, err)
+	}
+
+	errs = append(errs, doc.Validate()...)
```
Comment (mirogta) [inline pkg/register/documentBuilder.go:71]: 👍

**Review by mirogta — APPROVED** [review]
LGTM

File: pkg/register/document.go [hunk 4]
```diff
@@ -54,12 +69,120 @@ type RegisterDocument struct {
...
+		errs = append(errs, fmt.Errorf("metadata url is longer than max %d", maxURLLength))
+	}
+
+	// Check key/delegation names are unique in the document
+	bufOfNames := ""
+
+	for _, publicKey := range document.PublicKeys {
+		errs = append(errs, validatePublicKey(PublicKeyTypeString, publicKey)...)
+
+		if strings.Contains(bufOfNames, publicKey.ID+"|") {
```
Comment (StephanieBracaloni) [inline pkg/register/document.go:130]: To check the unicity we could use a `map[string]struc{}`, the operation on string concatenation and lookup are usually quite expensive. 
(Not the best example case ever but it shows the diff with a poor granularity :
with strings: https://ideone.com/Bt41zr
with map as a set: https://ideone.com/wZVZRl)

Comment (timchurchard) [reply pkg/register/document.go:130]: https://github.com/Iotic-Labs/iotics-identity-go/pull/9

**Pull request merged**